Modal logic (#1146)
UniMath/agda-unimath: src/modal-logic/deduction.lagda.md

> # Modal logic deduction

Reviewer: ```suggestion
# Deductions in modal logic
```

I would make the first word of this title "Deductions" so that it starts with the D, because the file titles will be listed by the file names in alphabetic ordering.

open import foundation.unit-type
open import foundation.universe-levels

open import foundation-core.cartesian-product-types

open import modal-logic.formulas
```

</details>

## Idea

The deduction system of modal logic is defined inductively as follows:

a ∈ T ------ (AX) T ⊢ a

T ⊢ a → b T ⊢ a ------------------ (MP) T ⊢ b

T ⊢ a ------ (NEC) T ⊢ □a

where `T` is a set of formulas and `a`, `b` are formulas.

Modal logic generated by a set of axioms is the smallest set of formulas that
contains the axioms and is closed under the deduction system.

## Definition

### Theory of modal formulas

```agda
module _
  {l1 : Level} (l2 : Level) (i : Set l1)
  where

  modal-theory : UU (l1 ⊔ lsuc l2)
  modal-theory = subtype l2 (modal-formula i)
```

### Deduction system

```agda
module _
  {l1 l2 : Level} {i : Set l1}
  where

  infix 5 _⊢ₘ_

  data _⊢ₘ_ (axioms : modal-theory l2 i) : modal-formula i → UU (l1 ⊔ l2) where
    modal-ax : {a : modal-formula i} → is-in-subtype axioms a → axioms ⊢ₘ a
    modal-mp :
      {a b : modal-formula i} → axioms ⊢ₘ a →ₘ b → axioms ⊢ₘ a → axioms ⊢ₘ b
    modal-nec : {a : modal-formula i} → axioms ⊢ₘ a → axioms ⊢ₘ □ₘ a
```

> ### Closure of a theory under the deduction system
> 
> ```agda
>   modal-logic-closure : modal-theory l2 i → modal-theory (l1 ⊔ l2) i
>   modal-logic-closure axioms a = trunc-Prop (axioms ⊢ₘ a)
> ```

Reviewer: We try to avoid modules that cross over sections, just because it is harder for the reader to keep track of what is assumed and sections seem a natural breaking point. This approach requires the arguments to be given a few more times, but we find it improves readability.

It also improves maintainability to break modules more often, because it is easier to move code around if it's not in a module that spans a large portion of the file.

````suggestion
### The deductive closure of a theory

A modal formula is in the deductive closure of a theory if there [exists](foundation.existence.md) a deduction of the formula from the theory.

```agda
module _
  {l1 l2 : Level} {i : Set l1}
  where
  
  modal-logic-closure : modal-theory l2 i → modal-theory (l1 ⊔ l2) i
  modal-logic-closure axioms a = trunc-Prop (axioms ⊢ₘ a)
```
````

Reviewer: I would also suggest:
* Renaming `modal-logic-closure` to `deductive-closure-modal-theory`
* Adding an entry `is-in-deductive-closure-modal-theory` instead of using `is-in-subtype (modal-logic-closure theory)` later on.

> ### Modal logic predicate
> 

Reviewer: A sentence explaining this definition informally would be nice. For instance:

```suggestion
### The predicate of being a modal logic

A {{#concept "modal logic" Agda=is-modal-logic}} is a theory of modal logic that contains its own deductive closure.
```

```agda
  is-modal-logic-Prop : modal-theory l2 i → Prop (l1 ⊔ l2)
  is-modal-logic-Prop theory =
    leq-prop-subtype (modal-logic-closure theory) theory

  is-modal-logic : modal-theory l2 i → UU (l1 ⊔ l2)
  is-modal-logic = type-Prop ∘ is-modal-logic-Prop
```

> ### TODO: Title

Reviewer: ```suggestion
### Any modal theory is contained in its own deductive closure
```

```agda

>   is-in-modal-logic-closure-deduction :

Reviewer: If my previous renaming suggestions are accepted, then this one could be named `is-in-deductive-closure-is-in-modal-theory`.

    {axioms : modal-theory l2 i} {a : modal-formula i} →
    axioms ⊢ₘ a → is-in-subtype (modal-logic-closure axioms) a
  is-in-modal-logic-closure-deduction = unit-trunc-Prop

> 
>   is-contradictory-modal-logic-Prop : modal-theory l2 i → Prop l2
>   is-contradictory-modal-logic-Prop logic = logic ⊥ₘ
> 
>   is-contradictory-modal-logic : modal-theory l2 i → UU l2
>   is-contradictory-modal-logic = type-Prop ∘ is-contradictory-modal-logic-Prop

Reviewer: I would name this `is-inconsistent-modal-theory-Prop`, and the next one `is-inconsistent-modal-theory`.

The point is that it is a predicate on modal theories, which isn't currently reflected in the name. I personally find `inconsistent` reading a bit better than `contradictory`, and it is also more consistent with your use of `consistent` of the name for consistent modal theories.

I would also put this in a new section.

One question I have about your definition of contradictory modal logics is that currently it says that a _theory_ is contradictory if it contains falsehood. However, theories are not deductively closed, so wouldn't it be more natural to say that modal theories are inconsistent if their deductive closure contains falsehood?

````suggestion
```

### Inconsistent modal theories

```agda
  is-inconsistent-modal-theory-Prop : modal-theory l2 i → Prop l2
  is-inconsistent-modal-theory-Prop logic = logic ⊥ₘ

  is-inconsistent-modal-theory : modal-theory l2 i → UU l2
  is-inconsistent-modal-theory = type-Prop ∘ is-contradictory-modal-logic-Prop
````

Author: > However, theories are not deductively closed, so wouldn't it be more natural to say that modal theories are inconsistent if their deductive closure contains falsehood?

On the one hand, this would be correct from a definitional point of view. But on the other hand, in proofs this definition is already used in relation to closures, therefore it is convenient to use it in this form. I'll think about how best to determine

  is-consistent-modal-logic-Prop : modal-theory l2 i → Prop l2
  is-consistent-modal-logic-Prop = neg-Prop ∘ is-contradictory-modal-logic-Prop

  is-consistent-modal-logic : modal-theory l2 i → UU l2

>   is-consistent-modal-logic = type-Prop ∘ is-consistent-modal-logic-Prop

Reviewer: It would make sense if the consistent and inconsistent modal theories each get their own sections `### Inconsistent modal theories` and `### Consistent modal theories`.
